deployers/whiskclient.go

```diff
@@ -20,194 +20,220 @@ package deployers
+	if len(credential.Value) == 0 || len(namespace.Value) == 0 || len(apiHost.Value) == 0 {
+		if utils.FileExists(manifestPath) {
+			mm := parsers.NewYAMLParser()
+			manifest, _ := mm.ParseManifest(manifestPath)
+			credential = GetPropertyValue(credential, manifest.Package.Credential, path.Base(manifestPath))
+			namespace = GetPropertyValue(namespace, manifest.Package.Namespace, path.Base(manifestPath))
+			apiHost = GetPropertyValue(apiHost, manifest.Package.ApiHost, path.Base(manifestPath))
+		}
+	}
+
+	// Third, we need to look up the variables in .wskprops file.
+	pi := whisk.PropertiesImp{
+		OsPackage: whisk.OSPackageImp{},
+	}
+
+	// The error raised here can be neglected, because we will handle it in the end of this function.
+	wskprops, _ := GetWskPropFromWskprops(pi, proppath)
+	credential = GetPropertyValue(credential, wskprops.AuthKey, WSKPROPS)
+	namespace = GetPropertyValue(namespace, wskprops.Namespace, WSKPROPS)
+	apiHost = GetPropertyValue(apiHost, wskprops.APIHost, WSKPROPS)
+
+	// now, read credentials from whisk.properties but this is only acceptable within Travis
+	// whisk.properties will soon be deprecated and should not be used for any production deployment
+	whiskproperty, _ := GetWskPropFromWhiskProperty(pi)
+	credential = GetPropertyValue(credential, whiskproperty.AuthKey, WHISKPROPERTY)
+	if credential.Source == WHISKPROPERTY {
+		fmt.Println("WARNING: The authentication key was retrieved from whisk.properties " +
+			"which will soon be deprecated please do not use it outside of Travis builds.")
+	}
+	namespace = GetPropertyValue(namespace, whiskproperty.Namespace, WHISKPROPERTY)
+	if namespace.Source == WHISKPROPERTY {
+		fmt.Println("WARNING: The namespace was retrieved from whisk.properties " +
+			"which will soon be deprecated please do not use it outside of Travis builds.")
+	}
+	apiHost = GetPropertyValue(apiHost, whiskproperty.APIHost, WHISKPROPERTY)
+	if apiHost.Source == WHISKPROPERTY {
+		fmt.Println("WARNING: The API host was retrieved from whisk.properties " +
+			"which will soon be deprecated please do not use it outside of Travis builds.")
+	}
+
+	// set namespace to default namespace if not yet found
+	if len(apiHost.Value) != 0 && len(credential.Value) != 0 && len(namespace.Value) == 0 {
+		namespace.Value = whisk.DEFAULT_NAMESPACE
+		namespace.Source = DEFAULTVALUE
+	}
+
+	// If we still can not find the values we need, check if it is interactive mode.
+	// If so, we prompt users for the input.
+	// The namespace is set to a default value at this point if not provided.
+	if len(apiHost.Value) == 0 && isInteractive == true {
+		host, err := promptForValue("\nPlease provide the hostname for OpenWhisk [default value is openwhisk.ng.bluemix.net]: ")
+		utils.Check(err)
+		if host == "" {
+			host = "openwhisk.ng.bluemix.net"
+		}
+		apiHost.Value = host
+		apiHost.Source = INTERINPUT
+	}
+
+	if len(credential.Value) == 0 && isInteractive == true {
```

Comment:
@pritidesai it seems we have many places where we use the "len" function to test for empty string; however, it seems GoLang convention is to simple test is  ```if mystring == "" ```.  Is there some downside to testing against a string literal (which should be faster)?

Reply:
Both `len(string) == 0` and `string == ""` are idiomatic ways of checking if string is empty. To find which option is faster and results showed `len()` is faster (for fun :) as the length of a string is stored in the variable itself during declaration.

```
package main

import (
    "fmt"
    "time"
    "github.com/davecgh/go-spew/spew"
)

func main() {

    // Length is 3.
    string1 := "cat"
    // Length is 13 times 5 which is 65.
    string2 := "cat0123456789" +
        "cat0123456789" +
        "cat0123456789" +
        "cat0123456789" +
        "cat0123456789"

    spew.Dump(string1)
    spew.Dump(string2)

    t0 := time.Now()

    // Version 1: length of short string.
    for i := 0; i < 10000000; i++ {
        result := len(string1)
        if result != 3 {
            return
        }
    }

    t1 := time.Now()

    // Version 2: length of long string.
    for i := 0; i < 10000000; i++ {
        result := len(string2)
        if result != 65 {
            return
        }
    }

    t2 := time.Now()
    // Print results.
    fmt.Println(t1.Sub(t0))
    fmt.Println(t2.Sub(t1))
}
``` 

Results:

$ go run  main.go
(string) (len=3) "cat"
(string) (len=65) "cat0123456789cat0123456789cat0123456789cat0123456789cat0123456789"
2.752905ms
2.761132ms

```
package main

import (
    "fmt"
    "time"
)

func main() {

    // Length is 3.
    string1 := "cat"
    // Length is 13 times 5 which is 65.
    string2 := "cat0123456789" +
        "cat0123456789" +
        "cat0123456789" +
        "cat0123456789" +
        "cat0123456789"

    t0 := time.Now()

    // Version 1: length of short string.
    for i := 0; i < 10000000; i++ {
        if string1 == "" {
            return
        }
    }

    t1 := time.Now()

    // Version 2: length of long string.
    for i := 0; i < 10000000; i++ {
        if string2 == "" {
            return
        }
    }

    t2 := time.Now()
    // Print results.
    fmt.Println(t1.Sub(t0))
    fmt.Println(t2.Sub(t1))
}
```

Results:

go run  main.go
2.91229ms
3.12793ms

Reply:
@pritidesai Nice!  IMO, performance is the end argument. Was just checking as I have seen on StackOverflow that the "voted" / preferred way is using the string literal compare on more than one thread.  Did not want some GoLang(ista?) person coming and criticizing our code...